### PR TITLE
Add ArrayIndices type

### PR DESCRIPTION
**Pull request by lucasteles**
Implements part of #670

### CODE REVIEW

File: source/array-index.d.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/**
+Extract the keys of an array or tuple.
```
Comment (sindresorhus) [inline source/array-index.d.ts:2]: The description here should be the same as in the readme.

I also think "keys" may be slightly confusing.

File: source/array-index.d.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+Use-case: Deal with valid keys for a const array.
+
+@example
+```
+import type {ArrayIndex} from 'type-fest';
+
+const weekDays = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday' ] as const;
```
Comment (sindresorhus) [inline source/array-index.d.ts:10]:
```suggestion
const weekdays = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'] as const;
```


weekdays is a word

File: source/array-index.d.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+type WeekDay = ArrayIndex<typeof weekDays>;
+
+const getWeekDayName = (day: WeekDay): WeekDayName => weekDays[day];
```
Comment (sindresorhus) [inline source/array-index.d.ts:14]: This is missing the `WeekDayName` type, and the example is kinda incomplete without a `ArrayValue` type.

**Comment by sindresorhus** [comment]
Should the name be plural? `ArrayIndices`

File: readme.md [hunk 4]
```diff
@@ -172,6 +172,7 @@ Click the type names for complete docs.
 - [`IsEqual`](source/is-equal.d.ts) - Returns a boolean for whether the two given types are equal.
 - [`TaggedUnion`](source/tagged-union.d.ts) - Create a union of types that share a common discriminant property.
 - [`IntRange`](source/int-range.d.ts) - Generate a union of numbers.
+- [`ArrayIndex`](source/array-index.d.ts) - Extract keys of an array.
```
Comment (sindresorhus) [inline readme.md:175]:
```suggestion
- [`ArrayIndex`](source/array-index.d.ts) - Provides valid indices for a constant array or tuple.
```

Maybe clearer?

File: source/array-index.d.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/**
+Extract the keys of an array or tuple.
+
+Use-case: Deal with valid keys for a const array.
```
Comment (sindresorhus) [inline source/array-index.d.ts:4]:
```suggestion
Use-case: This type is useful when working with constant arrays or tuples and you want to enforce type-safety for accessing elements by their indices.
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/array-index.d.ts (hunk 1 above)
Comment (lucasteles) [inline source/array-index.d.ts:2]:
```suggestion
Provides valid indices for a constant array or tuple.
```

File: source/array-index.d.ts (hunk 3 above)
Comment (lucasteles) [reply source/array-index.d.ts:14]: Agree, the PR for `ArrayValue` looks stoled, I can create another for it to merge together

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by lucasteles** [comment]
@sindresorhus fixed

**Pull request merged**